### DISCUSSION TITLE
Extra embedded system examples outside of GPS tutorial

### DISCUSSION THREAD

**Discussion by cawald-gtri** [nasa/fprime#763] [Q&A]
I just started learning fprime and I've been talking with @SterlingPeet about the basics. I'm an electrical engineer so reading through the gps example helped me because it ties the concepts into a project with physical hardware. Outside of that, I've been struggling to find other projects that would help guide me through learning f prime. The MathComponent tutorial was good but it didn't help me understand how to use fprime with hardware. Sterling mentioned that at the f prime workshop given at JPL/GT, there was a hands-on robotic arm project that was built using fprime. I found this repo https://github.com/fprime-community/fprime-course-materials that had the presentation but it didn't really go into depth on how to actually start coding the project. Are there any resources out there for the robot arm project or other projects with embedded hardware that walk through the coding a good bit?

**Comment by rawinza555** [comment]
There's a RPI demo that will kind of help you understanding how to use gpio or spi but not that much of documentation. I agree that there needed to be better documentation on these plus i2c. 

I'm actually interested to revamp the tutorial and add more stuff into it. Something similar to "F' for dummies" so that the beginners can understand easily. But I'm still a new learner myself and still learn how to use the available driver and stuff. I will see what I can do once I feel comfortable with this.

**Reply by Joshua-Anderson** [reply]
More tutorials and examples are always appreciated!

**Comment by Joshua-Anderson** [comment]
The good news is that embedded hardware isn't too different in F' than it would be elsewhere, so you can definitely use other more general resources to learn how to integrate hardware. The trick with F' is you usually wind up wrapping the platform specific hardware peripheral APIs with a F' component (See the F' Linux SPI and Linux I2C drivers as examples). A lot of this is platform specific so making sure you understand how to integrate the hardware with your target platform is a good first step.

I agree that a driver tutorial could be really useful... maybe a tutorial for how to use a common I2C sensor with F' on the raspberry pi?
We also don't have a large open source driver collection, so there aren't a lot of examples to go off.

**Reply by rawinza555** [reply]
Perhaps some sort of documents / comments on the componentimpl.cpp that introduce each driver, the ports and what need to be passed in between would be a good short term solution.

**Comment by LeStarch** [comment]
The GPS tutorial was intended to start showing how to work with embedded systems.  It shows how to work with a driver (serial), and convert some output into telemetry.

Right now it is "under construction" as it hasn't been moved to v2.0.0.

**Reply by LeStarch** [reply]
Reading not just the previous comment, I see the GPS tutorial was already looked at.

**Comment by LeStarch** [comment]
Is the desire here to see more F´, dive deeper into embedded systems, see more drivers, or something else?  Another way to ask the question is: what is the GPS lacking in representing an "embedded systems" tutorial?

**Reply by SterlingPeet** [reply]
@LeStarch I think the post here is born out of the frustration of not seeing a modern, up-to-date, fully packaged example that uses hardware.  Something where there is a parts list that can be purchased and reasonably expect the scaffolding to be in place to yield a complete example hardware project over relatively short learning timeframe.

I suspect the the lack thereof is largely bound to the fallout around releasing 2.0.0 without all of the examples being fully up to date with the new release.

**Reply by SterlingPeet** [reply]
Of course, I say this knowing full well that I have a reference mission sitting on my shelf in my office.  It is intended for this purpose, yet I have not found the time to write a tutorial for that project either.

**Reply by cawald-gtri** [reply]
My thoughts exactly. I would like a tutorial that is similar to the Math Component one that starts right after the installation steps, but works through a hardware project that involves the typical beginner embedded projects. The robot arm seemed to be a good example because it combined input with an output that used i2c and interacted with physical components. It also seems like it would have a good scaffolding to slowly increase the complexity of the project.

**Reply by rawinza555** [reply]
I was on the same boat as you @cawald-gtri. Took me a while to find out about these available driver and then once I know what and where I need to look for stuff then everything is getting better.

**Reply by capsulecorplab** [reply]
Have there been any updates on this topic of an "embedded systems" tutorial? I've been perusing the fprime discussions and community projects for insights into how to go about developing components that use the I2C driver and it appears there was an [I2C tutorial](https://github.com/nasa/fprime/discussions/810) developed at one point, but it seems to have disappeared from the interwebs. The IMU used on the system reference and baremetal reference is at least a working example, but it's not entirely clear how to go about looking at a datasheet for a sensor and creating a component for it. Anxious to hear anyone's thoughts on the subject.

**Comment by SterlingPeet** [comment]
I might be able to take a stab at recreating the robot arm project in 2.0.0 next week, and see where that gets.

Sterling

Sent from my mobile.

On Jun 24, 2021, at 12:25 PM, cawald-gtri ***@***.***> wrote:

﻿

My thoughts exactly. I would like a tutorial that is similar to the Math Component one that starts right after the installation steps, but works through a hardware project that involves the typical beginner embedded projects. The robot arm seemed to be a good example because it combined input with an output that used i2c and interacted with physical components. It also seems like it would have a good scaffolding to slowly increase the complexity of the project.

**Reply by cawald-gtri** [reply]
If you wouldn't mind me sitting there while you work through that, I could write up a tutorial as we go along which hopefully could be published as another learning guide.

**Reply by SterlingPeet** [reply]
yep, I'll give you a heads up when I'm working on it.

**Reply by LeStarch** [reply]
Keep me in-the-loop too, especially if there is something we can do to help this effort.

**Reply by knapsea** [reply]
We particularly like the concept of developing on a desktop as much as possible ( while  waiting for hardware design to be finalized ) , so would hope the cross-development aspects of FPrime be accentuated in the docs.

Thanks 
G
